server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -74,28 +74,50 @@ public class ThreadPool implements ReportingService<ThreadPoolInfo>, Scheduler,
+         * risks deadlock if there's a chance that the completion of the future depends on work being done on this pool. Unfortunately
+         * that's pretty likely in most cases because of how often this pool is used; it's really rare because of the high limit on the pool
+         * size, but when it happens  it is extremely harmful to the node.
+         * <p>
+         * This pool is also used for recovery-related work. The recovery subsystem bounds its own concurrency, and therefore the amount of
+         * recovery work done on the {@code #GENERIC} pool, via {@code cluster.routing.allocation.node_concurrent_recoveries} and related
+         * settings.
+         * <p>
+         * This pool does not reject any task. If you submit a task after the pool starts to shut down, it may simply never run.
          */
         public static final String GENERIC = "generic";
+
         /**
-         * Important management tasks that keep the cluster from falling apart.
-         * This thread pool ensures cluster coordination tasks do not get blocked by less critical tasks and can continue to make progress.
-         * This thread pool also defaults to a single thread, reducing contention on the Coordinator mutex.
+         * A thread pool solely for the use of the cluster coordination subsystem that relates to cluster state updates, master elections,
+         * cluster membership and so on.
+         * <p>
+         * This pool defaults to a single thread to avoid contention on {@code Coordinator#mutex}.
          */
         public static final String CLUSTER_COORDINATION = "cluster_coordination";
+
         public static final String GET = "get";
         public static final String ANALYZE = "analyze";
         public static final String WRITE = "write";
         public static final String SEARCH = "search";
         public static final String SEARCH_COORDINATION = "search_coordination";
         public static final String AUTO_COMPLETE = "auto_complete";
         public static final String SEARCH_THROTTLED = "search_throttled";
+
         /**
-         * Cluster management tasks. Tasks that manage data, and tasks that report on cluster health via statistics etc.
-         * Not a latency sensitive thread pool: some tasks may time be long-running; and the thread pool size is limited / relatively small.
+         * A thread pool for running tasks related to cluster management, including collecting and exposing stats in APIs and certain other
+         * internal tasks.
+         * <p>
+         * This pool is deliberately small in order to throttle the rate at which such tasks are executed and avoid diverting resources away
+         * from production-critical work such as indexing and search. You may run long-running (CPU-bound or IO-bound) tasks on this pool,
+         * but if the work relates to a REST API call then it must be cancellable in order to prevent an overexcited client from blocking or
+         * delaying other management work.
```

Comment:
Nice update. It does make me wonder about the GENERIC thread pool's high thread count, though. Does this imply that GENERIC work is more important for the system than MANAGEMENT work?

Reply:
I don't think it's anything so deliberate, it's just that we need a lot of threads because we don't really know what the pool is being used for, and there's an expectation that tasks submitted here will run almost straight away.

Reply:
> an expectation that tasks submitted here will run almost straight away.

... even though of course this makes no sense when under load.